File: ts/lex.ts
import type { Extensions } from "./main"

const lineTerminators = [0x0a, 0x0b, 0x0c, 0x0d, 0x85, 0x2028, 0x2029]

const isLineTerminator = (r: number): boolean => lineTerminators.includes(r)

const whitespaceRegex = /\s/

// all unicode chars with whitespace property
const isWhitespace = (r: number): boolean =>
	!isLineTerminator(r) && whitespaceRegex.test(String.fromCharCode(r))

const isControl = (r: number): boolean =>
	(r <= 0x1f || (r >= 0x7f && r <= 0x9f)) &&
	!isLineTerminator(r) &&
	!whitespaceRegex.test(String.fromCharCode(r))

// makes up for js bad unicode handling lel
const isHighSurrogate = (r: number): boolean => r >= 0xd800 && r <= 0xdbff
const isLowSurrogate = (r: number): boolean => r >= 0xdc00 && r <= 0xdfff

// characters not in any Unicode category
const isUnassigned = (r: number): boolean => r >= 0x40000 && r <= 0xeffff

// surrogate, private use, unassigned
const isForbidden = (r: number): boolean =>
	isControl(r) || isHighSurrogate(r) || r > 0x10ffff || isUnassigned(r)

// " # ; { }
const reserved = [0x22, 0x23, 0x3b, 0x7b, 0x7d]

const isReserved = (r: number, exts: Extensions): boolean =>
	reserved.includes(r) ||
	("ExpressionArguments" in exts && r === 0x28) /* ( */

const errForbidden = "illegal character"

class stream {
	src: string
	pos = 0

	constructor(src: string) {
		this.src = src
	}

	reading(): boolean {
		return this.pos < this.src.length
	}

	current(): [number, string] {
		// TODO: wip error type
		const r = this.src.charCodeAt(this.pos)
		if (Number.isNaN(r)) return [0, "EOF"]

		// if high surrogate, add second part (low surrogate)
		if (isHighSurrogate(r)) {
			const next = this.src.charCodeAt(this.pos + 1)
			if (!isLowSurrogate(next))
				return [0, `${errForbidden} U+${r.toString(16).toUpperCase()}`]

			this.pos += 1
			return [0x10000 + ((r - 0xd800) << 10) + (next - 0xdc00), ""]
		}

		if (isForbidden(r)) {
			// get illegal character as U+XXXX
			if (r < 0x10000)
				return [
					0,
					`${errForbidden} U+${r.toString(16).toUpperCase().padStart(4, "0")}`,
				]
			return [0, `${errForbidden} U+${r.toString(16).toUpperCase()}`]
		}

		return [r, ""]
	}

	increment(n: number) {
		this.pos += n
	}

	next(n: number): number {
		return this.src[this.pos + n]?.charCodeAt(0) || 0
	}
}

export type tokenType =
	| "Unicode"
	| "0qArgument"
	| "1qArgument"
	| "3qArgument"
	| "Newline"
	| "LineContinuation"
	| "Whitespace"
	| "Comment"
	| "Semicolon"
	| "OpenBrace"
	| "CloseBrace"

export type token = {
	Type: tokenType
	Content?: number[]
	Og?: number[]
}

// A directive “argument” shall be a sequence of one or more characters from the argument character set. The argument character set shall consist of any Unicode scalar value excluding characters from the white space, line terminator, reserved punctuator, and forbidden character sets.
const argumentOk = (r: number, exts: Extensions): boolean =>
	!isWhitespace(r) && !isLineTerminator(r) && !isReserved(r, exts)
const quotedArgumentOk = (r: number): boolean =>
	!isLineTerminator(r) && r !== 0x22 // "
const tripleQuotedArgumentOk = (r: number): boolean => r !== 0x22 // "

const errIncompleteEscape = new Error("incomplete escape sequence")
const errIllegalEscape = new Error("illegal escape character")
const errUnclosedQuoted = new Error("unclosed quoted")

function checkEscape(s: stream, r: number, quoted: number): [number, boolean] {
	if (r !== 0x5c /* \ */) return [r, false]

	s.increment(1)
	const [c, err] = s.current()
	if (err) {
		// find if error message starts with errForbidden
		if (err.startsWith(errForbidden) || quoted === 0) throw errIllegalEscape
		throw errIncompleteEscape
	}
	if (isWhitespace(c) || isLineTerminator(c)) {
		if (quoted === 3) {
			if (isLineTerminator(c)) throw errIncompleteEscape
			throw errIllegalEscape
		}
		if (quoted === 0 || (quoted === 1 && !isLineTerminator(c)))
			throw errIllegalEscape
		return [0, true] // r = 0 used to signify line terminatosr
	}

	return [c, true]
}

function getPunctuator(s: stream, pstr: string): number {
	let ps = pstr
	ps = ps.replaceAll("\r\n", "\n")
	ps = ps.replaceAll("\r", "\n")
	ps = ps.trim()

	const puncts = ps.split("\n")
	// sort puncts by length descending
	puncts.sort((a, b) => b.length - a.length)

	for (const p of puncts) {
		const rest = s.src.slice(s.pos)
		const l = p.length
		if (l <= rest.length && rest.slice(0, l) === p) {
			console.log("punctuator", p.charCodeAt(0), l)
			return l}
	}

	return 0
}

function lex0qArgument(s: stream, exts: Extensions): [number[], number[]] {
	const arg: number[] = []
	const ogarg: number[] = []

	while (s.reading()) {
		const [c, err] = s.current()
		if (err) throw new Error(err)
		if (
			!argumentOk(c, exts) ||
			(exts.PunctuatorArguments !== undefined &&
				getPunctuator(s, exts.PunctuatorArguments) !== 0)
		)
			return [arg, ogarg]

		const [ec, escd] = checkEscape(s, c, 0)
		if (escd) ogarg.push(0x5c /* \ */)

		arg.push(ec)
		ogarg.push(ec)
		s.increment(1)
	}

	return [arg, ogarg]
}

function lex1qArgument(s: stream): [number[], number[]] {
	const arg: number[] = []
	const ogarg: number[] = []

	while (s.reading()) {
		const [c, err] = s.current()
		if (err.startsWith(errForbidden)) throw new Error(errForbidden)
		if (!quotedArgumentOk(c)) {
			if (c !== 0x22 /* " */) throw errUnclosedQuoted

			s.increment(1)
			return [arg, ogarg]
		}

		const [ec, escd] = checkEscape(s, c, 1)
		if (escd) ogarg.push(0x5c /* \ */)

		if (ec === 0) {
			// escaped line terminators allowed in quoted arguments
			const [nc, _] = s.current()
			ogarg.push(nc)
			continue
		}

		arg.push(ec)
		ogarg.push(ec)
		s.increment(1)
	}

	throw errUnclosedQuoted
}

function lex3qArgument(s: stream): [number[], number[]] {
	const arg: number[] = []
	const ogarg: number[] = []

	let endsMatched = 0
	while (s.reading()) {
		const [c, err] = s.current()
		if (err.startsWith(errForbidden)) throw new Error(errForbidden)
		if (!tripleQuotedArgumentOk(c)) {
			if (c !== 0x22 /* " */) throw errUnclosedQuoted

			ogarg.push(c)
			s.increment(1)

			if (endsMatched === 2) return [arg, ogarg.slice(0, -3)]
			endsMatched++
			continue
		}
		if (endsMatched > 0) {
			for (let i = 0; i < endsMatched; i++) arg.push(0x22 /* " */)
			endsMatched = 0
			continue
		}

		const [ec, escd] = checkEscape(s, c, 3)
		if (escd) ogarg.push(0x5c /* \ */)

		arg.push(ec)
		ogarg.push(ec)
		s.increment(1)
	}

	throw errUnclosedQuoted
}

export function lex(input: string, exts: Extensions): token[] {
	if (input.includes("\ufffd"))
		// replacement character
		throw new Error("malformed UTF-8")

	let src = input
	const ts: token[] = []

	// remove BOMs
	if (src.startsWith("\ufeff")) {
		ts.push({ Type: "Unicode", Content: [0xfeff] })
		src = src.slice(1)
	} else if (src.startsWith("\ufffe")) {
		ts.push({ Type: "Unicode", Content: [0xfffe] })
		src = src.slice(1)
	}

	// remove ^Z
	let removeCtrlZ = false
	if (src.endsWith("\u001a")) {
		removeCtrlZ = true
		src = src.slice(0, -1)
	}

	// character buffer
	const buf: number[] = []
	for (let i = 0; i < src.length; i++) {
		const c = src.charCodeAt(i)
		buf.push(c)
	}

	// check for forbidden characters must be done based on token/location

	for (const s = new stream(src); s.reading(); ) {
		const [c, err] = s.current()
		if (err) throw new Error(err)

		const op = s.pos
		if (isLineTerminator(c)) {
			s.increment(1)
			ts.push({ Type: "Newline", Content: [c] })
		} else if (isWhitespace(c)) {
			s.increment(1)
			ts.push({ Type: "Whitespace", Content: [c] })
		} else if (
			"CStyleComments" in exts &&
			c === 0x2f &&
			s.next(1) === 0x2f //
		) {
			// C-style comment
			for (s.increment(1); ; ) {
				s.increment(1)
				const [c, err] = s.current()
				if (err.startsWith(errForbidden)) throw new Error(errForbidden)
				if (err || isLineTerminator(c)) break
			}
			const content = buf.slice(op + 2, s.pos)
			ts.push({
				Type: "Comment",
				Content: content,
				Og: Array.from(Uint8Array.from(`#${content}`)),
			})
		} else if (c === 0x23 /* # */) {
			// comment until end of line
			while (true) {
				s.increment(1)
				const [c, err] = s.current()
				if (err.startsWith(errForbidden)) throw new Error(errForbidden)
				if (err || isLineTerminator(c)) break
			}
			const content = buf.slice(op + 1, s.pos)
			ts.push({
				Type: "Comment",
				Content: content,
				Og: Array.from(Uint8Array.from(`#${content}`)),
			})
		} else if (
			"CStyleComments" in exts &&
			c === 0x2f &&
			s.next(1) === 0x2a /* */
		) {
			// block comment
			for (s.increment(1); ; ) {
				s.increment(1)
				const [c, err] = s.current()
				if (err.startsWith(errForbidden)) throw new Error(errForbidden)
				if (err) throw new Error("unterminated multi-line comment")
				if (c === 0x2a /* * */ && s.next(1) === 0x2f /* / */) break
			}
			const content = buf.slice(op + 2, s.pos)
			ts.push({
				Type: "Comment",
				Content: content,
				Og: Array.from(Uint8Array.from(`/*${content}*/`)),
			})
			s.increment(2) // */
		} else if (c === 0x3b /* ; */) {
			s.increment(1)
			ts.push({ Type: "Semicolon" })
		} else if (c === 0x7b /* { */) {
			s.increment(1)
			ts.push({ Type: "OpenBrace" })
		} else if (c === 0x7d /* } */) {
			s.increment(1)
			ts.push({ Type: "CloseBrace" })
		} else if (c === 0x5c /* \ */ && isLineTerminator(s.next(1))) {
			s.increment(2)
			ts.push({ Type: "LineContinuation" })
		} else if ("ExpressionArguments" in exts && c === 0x28 /* ( */) {
			// read until corresponding closing parenthesis
			let depth = 0
			while (true) {
				s.increment(1)
				const [c, err] = s.current()
				if (err.startsWith(errForbidden)) throw new Error(errForbidden)
				if (err || isLineTerminator(c))
					throw new Error("incomplete expression")
				if (c === 0x28 /* ( */) depth++
				else if (c === 0x29 /* ) */) {
					if (depth === 0) break
					depth--
				}
			}
			const content = buf.slice(op + 1, s.pos)
			ts.push({
				Type: "0qArgument",
				Content: content,
				Og: Array.from(Uint8Array.from(`(${content})`)),
			})
			s.increment(1) // )
		} else if (
			exts.PunctuatorArguments !== undefined &&
			getPunctuator(s, exts.PunctuatorArguments) !== 0
		) {
			// read punctuator as argument
			s.increment(getPunctuator(s, exts.PunctuatorArguments))
			const content = buf.slice(op, s.pos)

		console.log("punctuator", content)

			ts.push({ Type: "0qArgument", Content: content, Og: content })
		} else if (c === 0x22 && s.next(1) === 0x22 && s.next(2) === 0x22) {
			// triple quated argument
			s.increment(3)
			const [arg, ogarg] = lex3qArgument(s)
			ts.push({ Type: "3qArgument", Content: arg, Og: ogarg })
		} else if (c === 0x22) {
			// quoted argument
			s.increment(1)
			const [arg, ogarg] = lex1qArgument(s)
			ts.push({ Type: "1qArgument", Content: arg, Og: ogarg })
		} else {
			// unquoted argument
			const [arg, ogarg] = lex0qArgument(s, exts)
			ts.push({ Type: "0qArgument", Content: arg, Og: ogarg })
		}
	}

	if (removeCtrlZ) ts.push({ Type: "Unicode", Content: [0x001a] })

	return ts
}

File: go/lex.go
package confetti

import (
	"errors"
	"fmt"
	"slices"
	"strings"
	"unicode"
	"unicode/utf8"
)

var lineTerminators = []rune{'\r', '\n', '\u000a', '\u000b', '\u000c', '\u000d', '\u0085', '\u2028', '\u2029'}

func isLineTerminator(r rune) bool {
	return slices.Contains(lineTerminators, r)
}

// all unicode chars with whitespace property
func isWhitespace(r rune) bool {
	return !isLineTerminator(r) && unicode.IsSpace(r)
}

// characters not in any Unicode category
func isUnassigned(r rune) bool {
	return r >= 0x40000 && r <= 0xEFFFF
}

// surrogate, private use, unassigned
func isForbidden(r rune) bool {
	return unicode.Is(unicode.Cc, r) && !isLineTerminator(r) && !isWhitespace(r) || unicode.Is(unicode.Cs, r) || r > 0x10FFFF || isUnassigned(r)
}

var reserved = []rune{'"', '#', ';', '{', '}'}

func isReserved(r rune, exts Extensions) bool {
	return slices.Contains(reserved, r) || exts.Has(ExtExpressionArguments) && r == '('
}

type stream struct {
	src []rune
	pos int
}

func (s *stream) reading() bool {
	return s.pos < len(s.src)
}

var (
	errEOF       = errors.New("EOF")
	errForbidden = errors.New("illegal character")
)

func (s *stream) current() (c rune, err error) {
	if s.pos >= len(s.src) {
		// panic("EOF")
		return 0, errEOF
	} else if c = s.src[s.pos]; isForbidden(c) {
		// get illegal character as U+XXXX
		if c < 0x10000 {
			return 0, fmt.Errorf("%w U+%04X", errForbidden, c)
		}
		return 0, fmt.Errorf("%w U+%X", errForbidden, c)
	}

	return
}

func (s *stream) next(n int) rune {
	if s.pos+n < len(s.src) {
		return s.src[s.pos+n]
	}
	return 0
}

// tokens
type tokenType uint8

const (
	tokUnicode tokenType = iota
	tok0qArgument
	tok1qArgument
	tok3qArgument
	tokNewline
	tokLineContinuation
	tokWhitespace
	tokComment
	tokSemicolon
	tokOpenBrace
	tokCloseBrace
)

type token struct {
	Type        tokenType
	Content, Og string
}

// A directive “argument” shall be a sequence of one or more characters from the argument character set. The argument character set shall consist of any Unicode scalar value excluding characters from the white space, line terminator, reserved punctuator, and forbidden character sets.
func argumentOk(r rune, exts Extensions) bool {
	return !isWhitespace(r) && !isLineTerminator(r) && !isReserved(r, exts)
}

func quotedArgumentOk(r rune) bool {
	return !isLineTerminator(r) && r != '"'
}

func tripleQuotedArgumentOk(r rune) bool {
	return r != '"'
}

var (
	errIncompleteEscape = errors.New("incomplete escape sequence")
	errIllegalEscape    = errors.New("illegal escape character")
	errUnclosedQuoted   = errors.New("unclosed quoted")
)

func checkEscape(s *stream, c rune, quoted uint8) (r rune, escaped bool, err error) {
	if c != '\\' {
		return c, false, nil
	}

	s.pos++
	if c, err = s.current(); err != nil {
		if errors.Is(err, errForbidden) {
			return 0, false, errIllegalEscape
		} else if quoted > 0 {
			return 0, false, errIncompleteEscape
		}
		return 0, false, errIllegalEscape
	} else if isWhitespace(c) || isLineTerminator(c) {
		if quoted == 3 {
			if isLineTerminator(c) {
				return 0, false, errIncompleteEscape
			}
			return 0, false, errIllegalEscape
		} else if quoted == 0 || quoted == 1 && !isLineTerminator(c) {
			return 0, false, errIllegalEscape
		}
		return 0, true, nil // r = 0 used to signify line terminator
	}
	return c, true, nil
}

func getPunctuator(s *stream, ps string) (l int) {
	ps = strings.ReplaceAll(ps, "\r\n", "\n")
	ps = strings.ReplaceAll(ps, "\r", "\n")
	ps = strings.TrimSpace(ps)

	puncts := strings.Split(ps, "\n")
	// sort puncts by length descending
	slices.SortFunc(puncts, func(a, b string) int {
		return len([]rune(b)) - len([]rune(a)) // footgun: len() counts bytes for strings
	})
	// fmt.Printf("punctuators: %q\n", puncts)

	for _, p := range puncts {
		rest := s.src[s.pos:]

		rp := []rune(p)
		l = len(rp)
		// fmt.Printf("%q %d\n", p, l)
		if l <= len(rest) && string(rest[:l]) == p {
			// fmt.Printf("punctuator: %q\n", p)
			return
		}
	}
	// fmt.Printf("not a punctuator: %q\n", s.src[s.pos:])
	return 0
}

func lex0qArgument(s *stream, exts Extensions) (arg, ogarg []rune, err error) {
	for s.reading() {
		c, err := s.current()
		if err != nil {
			return nil, nil, err
		} else if !argumentOk(c, exts) || exts.Has(ExtPunctuatorArguments) && getPunctuator(s, exts[ExtPunctuatorArguments]) != 0 {
			return arg, ogarg, nil
		}

		ec, escd, err := checkEscape(s, c, 0)
		if err != nil {
			return nil, nil, err
		} else if escd {
			ogarg = append(ogarg, '\\')
		}

		arg = append(arg, ec)
		ogarg = append(ogarg, ec)
		s.pos++
	}

	return
}

func lex1qArgument(s *stream) (arg, ogarg []rune, err error) {
	for s.reading() {
		c, err := s.current()
		if errors.Is(err, errForbidden) {
			return nil, nil, errForbidden
		} else if !quotedArgumentOk(c) {
			if c != '"' {
				return nil, nil, errUnclosedQuoted
			}

			s.pos++
			return arg, ogarg, nil
		}

		ec, escd, err := checkEscape(s, c, 1)
		if err != nil {
			return nil, nil, err
		} else if escd {
			ogarg = append(ogarg, '\\')
		}

		if ec == 0 { // escaped line terminators allowed in quoted arguments
			nc, _ := s.current()
			ogarg = append(ogarg, nc)
		} else {
			arg = append(arg, ec)
			ogarg = append(ogarg, ec)
		}
		s.pos++
	}

	return nil, nil, errUnclosedQuoted
}

func lex3qArgument(s *stream) (arg, ogarg []rune, err error) {
	for endsMatched := 0; s.reading(); {
		c, err := s.current()
		if errors.Is(err, errForbidden) {
			return nil, nil, errForbidden
		} else if !tripleQuotedArgumentOk(c) {
			if c != '"' {
				return nil, nil, errUnclosedQuoted
			}

			ogarg = append(ogarg, c)
			s.pos++

			endsMatched++
			if endsMatched != 3 {
				continue
			}
			return arg, ogarg[:len(ogarg)-3], nil
		} else if endsMatched > 0 {
			arg = append(arg, slices.Repeat([]rune{'"'}, endsMatched)...)
			endsMatched = 0
			continue
		}

		ec, escd, err := checkEscape(s, c, 3)
		if err != nil {
			return nil, nil, err
		} else if escd {
			ogarg = append(ogarg, '\\')
		}

		arg = append(arg, ec)
		ogarg = append(ogarg, ec)
		s.pos++
	}

	return nil, nil, errUnclosedQuoted
}

func lex(src string, exts Extensions) (ts []token, err error) {
	// remove BOMs
	if strings.HasPrefix(src, "\uFEFF") {
		ts = append(ts, token{Type: tokUnicode, Content: "\uFEFF"})
		src = src[3:]
	} else if strings.HasPrefix(src, "\uFFFE") {
		ts = append(ts, token{Type: tokUnicode, Content: "\uFFFE"})
		src = src[3:]
	}

	// remove ^Z
	if strings.HasSuffix(src, "\u001a") {
		defer func() {
			ts = append(ts, token{Type: tokUnicode, Content: "\u001a"})
		}()
		src = src[:len(src)-1]
	}

	if !utf8.ValidString(src) {
		return nil, errors.New("malformed UTF-8")
	}

	// check for forbidden characters must be done based on token/location

	for s := (stream{src: []rune(src)}); s.reading(); {
		c, err := s.current()
		if err != nil {
			break
		}

		switch op := s.pos; {
		case isLineTerminator(c):
			s.pos++
			ts = append(ts, token{Type: tokNewline, Content: string(c)})

		case isWhitespace(c):
			s.pos++
			ts = append(ts, token{Type: tokWhitespace, Content: string(c)})

		case exts.Has(ExtCStyleComments) && c == '/' && s.next(1) == '/': // C-style comment
			for s.pos += 2; ; s.pos++ {
				if c, err = s.current(); errors.Is(err, errForbidden) {
					return nil, errForbidden
				} else if err != nil || isLineTerminator(c) {
					break
				}
			}
			content := string(s.src[op+2 : s.pos])
			ts = append(ts, token{Type: tokComment, Content: content, Og: "//" + content})

		case c == '#': // comment until end of line
			for s.pos++; ; s.pos++ {
				if c, err = s.current(); errors.Is(err, errForbidden) {
					return nil, errForbidden
				} else if err != nil || isLineTerminator(c) {
					break
				}
			}
			content := string(s.src[op+1 : s.pos])
			ts = append(ts, token{Type: tokComment, Content: content, Og: "#" + content})

		case exts.Has(ExtCStyleComments) && c == '/' && s.next(1) == '*': // block comment
			for s.pos += 2; ; s.pos++ {
				if c, err = s.current(); errors.Is(err, errForbidden) {
					return nil, errForbidden
				} else if err != nil {
					return nil, errors.New("unterminated multi-line comment")
				} else if c == '*' && s.next(1) == '/' {
					break
				}
			}
			content := string(s.src[op+2 : s.pos])
			ts = append(ts, token{Type: tokComment, Content: content, Og: "/*" + content + "*/"})
			s.pos += 2 // */

		case c == ';':
			s.pos++
			ts = append(ts, token{Type: tokSemicolon})

		case c == '{':
			s.pos++
			ts = append(ts, token{Type: tokOpenBrace})

		case c == '}':
			s.pos++
			ts = append(ts, token{Type: tokCloseBrace})

		case c == '\\' && isLineTerminator(s.next(1)):
			s.pos += 2
			ts = append(ts, token{Type: tokLineContinuation})

		case exts.Has(ExtExpressionArguments) && c == '(':
			s.pos++
			// read until corresponding closing parenthesis
			for depth := 0; ; s.pos++ {
				if c, err = s.current(); errors.Is(err, errForbidden) {
					return nil, errForbidden
				} else if err != nil || isLineTerminator(c) {
					return nil, errors.New("incomplete expression")
				} else if c == '(' {
					depth++
				} else if c == ')' {
					if depth == 0 {
						break
					}
					depth--
				}
			}
			content := string(s.src[op+1 : s.pos])
			ts = append(ts, token{Type: tok0qArgument, Content: content, Og: "(" + content + ")"})
			s.pos++

		case exts.Has(ExtPunctuatorArguments) && getPunctuator(&s, exts[ExtPunctuatorArguments]) != 0:
			// read punctuator as argument
			s.pos += getPunctuator(&s, exts[ExtPunctuatorArguments])
			content := string(s.src[op:s.pos])
			ts = append(ts, token{Type: tok0qArgument, Content: content, Og: content})

		case c == '"' && s.next(1) == '"' && s.next(2) == '"': // triple quoted argument
			s.pos += 3
			arg, ogarg, err := lex3qArgument(&s)
			if err != nil {
				return nil, err
			}
			ts = append(ts, token{Type: tok3qArgument, Content: string(arg), Og: string(ogarg)})

		case c == '"': // quoted argument
			s.pos++
			arg, ogarg, err := lex1qArgument(&s)
			if err != nil {
				return nil, err
			}
			ts = append(ts, token{Type: tok1qArgument, Content: string(arg), Og: string(ogarg)})

		default: // unquoted argument
			arg, ogarg, err := lex0qArgument(&s, exts)
			if err != nil {
				return nil, err
			}
			ts = append(ts, token{Type: tok0qArgument, Content: string(arg), Og: string(ogarg)})
		}
	}

	return
}
